\documentclass[twocolumn]{article}

% packages
\usepackage{amssymb} 
\usepackage{amsmath}
\usepackage{algorithm}
\usepackage{algpseudocode}

% commands
\newcommand{\set}[1]{\left\{#1\right\}}
\newcommand{\seq}[1]{\langle #1 \rangle}
\newcommand{\type}[1]{\mathrm{#1}}
\newcommand{\mat}[1]{\mathbf{#1}}

\newcommand{\fig}[1]{Fig.~#1}
\newcommand{\subfigure}[1]{\textbf{(#1)}}
\newcommand{\sect}[1]{Sec.~#1}
\newcommand{\tab}[1]{Table~#1}
\newcommand{\alg}[1]{Alg.~#1}
\newcommand{\lst}[1]{List.~#1}
\newcommand{\eqn}[1]{Eq.~#1}
\newcommand{\thm}[1]{Thm.~#1}
\newcommand{\lem}[1]{Lemma~#1}


% bibiliographic info
\title{Pattern formation by evolution of cellular automata}
\author{Alex Garruss \and Casey Grun}
\date{\today}

\begin{document}
\maketitle

% =============================================================================
\section{Introduction}

Elementary two-dimensional cellular automata are simple computational programs defined by a rule set and seeded with an initial condition state.  The rule set determines the current value of a cell given a local neighborhood of previously computed cell states.  The initial condition space can significantly impact the immediate computational context, drastically affect the overall state of the automata during computation, and ultimately determine the diversity of output states for a given fixed rule set and output definition.
A traditional computer programming paradigm is frequently ``top-down'' in the sense rational paradigms and requirements of exactness for results (top-level outcomes) are directly achieved by design of lower-level procedures.  In contrast, genetic programs as they exist through evolution tend to arise from the natural ``bottom-up'' paradigm, beginning with simple cellular interactions that form higher-order structures and patterns.
Cellular automata provide a computational paradigm to explore simple rules and resulting complexity.  Cellular automata states are abstract and allow granularity corresponding to any part of an object or process.  Rule 110 was shown (Cook, 2004) to be Turing complete by showing equivalence of the automata to cyclic tags, previously proven to be complete.  This feature of the rule set implies a capability to harbor a vast diversity of output configurations, and potentially capable of harboring structures or patterns within the computed landscape of the running automata.

Emerging studies over the past several decades have looked to various cellular automata to describe patterns in biology.  In 1999 the theoretical biologist Andreas Deutsch proposed a migration automata to model cellular movement and morphogenetic pattern formation.  After building and refining a model (although not an elementary CA as proposed in the project), he describes the process of inferring simulated outcomes from observing rule definitions themselves.  For our proposed work, we expect to learn information about selected automata rule sets or initial conditions after computational evolution.  Deutsch proposes further morphogenetic description possibilities via simple automata to investigate other areas of pattern formation such as body formation.  We agree with his proposition, noting the sequential, hierarchical, and spatially-local pattern formation of the fly embryo (Clyde et al. 2003).  Furthermore, Deutsch suggests extending the model to link proper pattern formation to subsequent cell decisions, such as differentiation.  This suggestion could be interpreted in our proposed framework to shift rule sets upon a resulting pattern of interest during computation.  Other groups (reviewed by Ermentrout et al. 1993) have also approached pattern formation, along with areas in neurobiology and population dynamics.  Our approach differs dramatically in most cases since most of the previous groups rationally target mathematical models for a task at hand, whereas we propose to use an elementary CA to evolve, at least preliminarily, the input conditions in an non-rational, unsupervised way. 
We will evaluate some of the previous work combining CA-GA in the final paper.

Our project is to use an evolutionary algorithm to explore the process of computational pattern formation using cellular automata. We take inspiration from Stephen Wolframâ€™s program to exhaustively explore the behavior of each possible rule set for an elementary cellular automaton, and would like to perform a similar such exploration of the space of initial conditions for such an automaton. Rather than exhaustively enumerate and then examine all possible initial conditions, we will take a restricted, evolutionary approach. Specifically, we will attempt to answer the question: can we use evolutionary algorithms to generate initial conditions (``genotypes'') for a cellular automaton that can produce arbitrary output patterns (``phenotypes'')? Our starting point is Rule 110, a rule set for an elementary cellular automaton that has been shown capable of simulating a Turing machine. 

There are several important challenges in addressing this question:
\begin{description}
\item[Formulation of the evolutionary process] --- It is not immediately obvious how to map cellular automata on to a genotype space which can be evolved; similarly, how should one perform recombination on the features of the genotype space? We explore several models (see Methods below) for the genotype and fitness evaluation of our cellular automata. 
\item[Discontinuity of fitness landscape] ---  A fundamental premise of evolutionary search algorithms is that there exists some fitness process that can be expressed as a function of the genotype of the objects under selection, and that this function is (mostly) smooth. That is, small changes in the genotype should not (usually) produce large discontinuous jumps. However, in our case, the function mapping genotype (initial conditions) to fitness is in principle not smooth---since Rule 110 can simulate a Turing machine, the function mapping initial conditions to output conditions (and in turn to the fitness) is undecidable (if the lifetime of the automaton is not fixed). Therefore, one motivation for this project is a search for a locally smooth region of this highly discontinuous fitness landscape---are there initial conditions such that small perturbations produce little variation in the output?
\item[Universal computation $\ne$ arbitrary patterns] --- It is not obvious to us what class of patterns can be formed by an automaton running Rule 110, given various initial conditions, or what modifications to the automaton may be necessary and sufficient to allow formation of arbitrary patterns. It does not seem likely that, just because Rule 110 is Turing universal, it can also produce arbitrary output patterns. It does seem possible that arbitrary pattern generation could be possible with, for instance, a fixed basis function transformation on the output vector. We have attempted to explore and understand the space of patterns that can be formed by these types of automata.
\end{description}


% =============================================================================
\section{Methods}

% -----------------------------------------------------------------------------
\subsection{Cellular Automata}
% TODO: Alex

% -----------------------------------------------------------------------------
\subsection{Evolutionary Algorithm}
We implemented an general evolutionary algorithm in C++ to select a pool of $n$ genomes according to fitness.  At each evolutionary iteration, the algorithm would: mutate all genomes, kill off (assign fitness of $-\infty$) any genomes older than a certain age, and evaluate the fitness of each genome. Then it would sort the genomes in order of decreasing fitness; the $m$ ``fittest'' genotypes would remain in the pool unchanged, while the bottom $n-m$ genotypes would be replaced by genotypes generated by recombining two random genotypes from the surviving pool. This process is summarized in \alg{\ref{alg:ga}}

\begin{algorithm}
\label{alg:ga}
\begin{algorithmic}[1]
	\Procedure{GeneticAlgorithm}{pool size $n$, generation size $m$, generations $s$}
	\State $P \gets \seq{\text{$n$ random genotypes $p_1 \ldots p_n$}}$
	\State $F \gets \seq{f_1 \ldots f_n}$
	\Statex \Comment Initialize gene pool, fitnesses
	\Repeat
		\ForAll{$i \in \set{1 \ldots n}$}
			\State Mutate gene $p_i$
			\If{$p_i$ older than threshold}
				\State $f_i \gets -\infty$
			\Else
				\State $f_i \gets$ fitness of $p_i$
			\EndIf
		\EndFor
		\State Sort gene pool $P$ by fitnesses $f_1 \ldots f_n$
		\ForAll{$i \in \set{m \ldots n}$}
			\State Choose two random genotypes 
			\State $j,k \in \set{1 \ldots m}$
			\State $p_i \gets$ Recombine $p_j, p_k$
		\EndFor
	\Until{Stopping condition is met}
	\EndProcedure
\end{algorithmic}
\end{algorithm}

In our representation, the genomes were each bit vectors of length $\ell$. Mutation occurs choosing $\phi*\ell$ bit(s) in a genome uniformly randomly (that is, some fraction $\phi$ of each genome will be subject to mutaton), then independently setting each of those bits to a random value, 0 or 1. Recombination occurs by choosing choosing some uniformly random crossover point $\ell' \in \set{1\ldots\ell}$, then concatenating the first $\ell'$ bits from one parent with the last $\ell-\ell'$ bits from the other parent.  

The fitness of each genome is evaluated by running a cellular automaton for a fixed number of time steps, with the genome sequence (the ``genotype'') as the initial state. The final state of the automaton is the ``phenotype.'' The phenotype is compared by Hamming distance with some target bit vector (also of length $\ell$); the fitness is given as the negative Hamming distance. 

% =============================================================================
\section{Results}

% -----------------------------------------------------------------------------
\subsection{Evolutionary Pattern Formation}

% -----------------------------------------------------------------------------
\subsection{Fitness Landscape Smoothness}

% =============================================================================
\section{Conclusion}


\end{document}